src/lib/util/QueryBuilder.js

```diff
@@ -0,0 +1,120 @@
+const { isObject, mergeDefault } = require('./util');
+const { DEFAULTS: { DATATYPES } } = require('./constants');
+const Type = require('./Type');
+
+class QueryBuilder {
+
+	/**
+	 * @typedef {Object} QueryBuilderDatatype
+	 * @property {Function} [array = () => 'TEXT'] A function that converts the current datatype into an array
+	 * @property {Function} [resolver = null] The resolver to convert JS objects into compatible data for the SQL database. This can be used to prevent SQL injections
+	 * @property {string} type The name of the datatype, e.g. VARCHAR, DATE, or BIT
+	 */
+
+	/**
+	 * @typedef {QueryBuilderDatatype} QueryBuilderOptions
+	 * @property {Function} [formatDatatype] The datatype formatter for the SQL database
+	 * @property {Function} [arrayResolver] The specialized resolver for array keys
+	 */
+
+	/**
+	 * @since 0.5.0
+	 * @param {Object<QueryBuilderDatatype>} datatypes The datatype to insert
+	 * @param {QueryBuilderOptions} [options = {}] The default options for all datatypes plus formatDatatype
+	 */
+	constructor(datatypes = {}, { array = () => 'TEXT', resolver = null, type = null, arrayResolver, formatDatatype } = {}) {
```

Comment:
`{ array = () => 'TEXT', resolver = null, type = null, arrayResolver, formatDatatype }` should be defaulted in constants and mergeDefault, instead of here.

Also, consider that all options in both datatypes and options are optional. This really should all be one parameter, options, mergeDefault, then `const { array, resolver, type, arrayResolver, formatDatatype, ...datatypes } = options;`